SU2_CFD/src/solvers/CEulerSolver.cpp

```diff
@@ -3325,99 +3300,149 @@ void CEulerSolver::Centered_Residual(CGeometry *geometry, CSolver **solver_conta
+  auto chunkSize = roundUpDiv(OMP_MIN_SIZE, ColorGroupSize)*ColorGroupSize;
+
+  /*--- Loop over edge colors. ---*/
+  for (auto color : EdgeColoring)
+  {
+  SU2_OMP_FOR_DYN(chunkSize)
+  for(auto k = 0ul; k < color.size; ++k) {
+
+    auto iEdge = color.indices[k];
+#else
+  /*--- Natural coloring. ---*/
+  {
+  for (auto iEdge = 0ul; iEdge < geometry->GetnEdge(); iEdge++) {
+#endif
+    unsigned short iDim, iVar;
 
     /*--- Points in edge and normal vectors ---*/
 
-    iPoint = geometry->edge[iEdge]->GetNode(0); jPoint = geometry->edge[iEdge]->GetNode(1);
+    auto iPoint = geometry->edge[iEdge]->GetNode(0);
+    auto jPoint = geometry->edge[iEdge]->GetNode(1);
+
     numerics->SetNormal(geometry->edge[iEdge]->GetNormal());
 
+    auto Coord_i = geometry->node[iPoint]->GetCoord();
+    auto Coord_j = geometry->node[jPoint]->GetCoord();
+
     /*--- Roe Turkel preconditioning ---*/
 
     if (roe_turkel) {
-      sqvel = 0.0;
+      su2double sqvel = 0.0;
       for (iDim = 0; iDim < nDim; iDim ++)
-        sqvel += config->GetVelocity_FreeStream()[iDim]*config->GetVelocity_FreeStream()[iDim];
+        sqvel += pow(config->GetVelocity_FreeStream()[iDim], 2);
       numerics->SetVelocity2_Inf(sqvel);
     }
 
     /*--- Grid movement ---*/
 
-    if (dynamic_grid)
-      numerics->SetGridVel(geometry->node[iPoint]->GetGridVel(), geometry->node[jPoint]->GetGridVel());
+    if (dynamic_grid) {
+      numerics->SetGridVel(geometry->node[iPoint]->GetGridVel(),
+                           geometry->node[jPoint]->GetGridVel());
+    }
 
-    /*--- Get primitive variables ---*/
+    /*--- Get primitive and secondary variables ---*/
 
-    V_i = nodes->GetPrimitive(iPoint); V_j = nodes->GetPrimitive(jPoint);
-    S_i = nodes->GetSecondary(iPoint); S_j = nodes->GetSecondary(jPoint);
+    auto V_i = nodes->GetPrimitive(iPoint); auto V_j = nodes->GetPrimitive(jPoint);
+    auto S_i = nodes->GetSecondary(iPoint); auto S_j = nodes->GetSecondary(jPoint);
 
-    /*--- High order reconstruction using MUSCL strategy ---*/
+    /*--- Set them with or without high order reconstruction using MUSCL strategy. ---*/
 
-    if (muscl) {
+    if (!muscl) {
 
-      for (iDim = 0; iDim < nDim; iDim++) {
-        Vector_i[iDim] = 0.5*(geometry->node[jPoint]->GetCoord(iDim) - geometry->node[iPoint]->GetCoord(iDim));
-        Vector_j[iDim] = 0.5*(geometry->node[iPoint]->GetCoord(iDim) - geometry->node[jPoint]->GetCoord(iDim));
-      }
+      numerics->SetPrimitive(V_i, V_j);
+      numerics->SetSecondary(S_i, S_j);
+
+    }
+    else {
+      /*--- Reconstruction ---*/
+
+      /*--- Static arrays of primitives and secondaries (thread safety). ---*/
 
-      Gradient_i = nodes->GetGradient_Reconstruction(iPoint);
-      Gradient_j = nodes->GetGradient_Reconstruction(jPoint);
+      su2double Primitive_i[MAXNVAR] = {0.0}, Primitive_j[MAXNVAR] = {0.0};
+      su2double Secondary_i[MAXNVAR] = {0.0}, Secondary_j[MAXNVAR] = {0.0};
```

Comment:
For efficiency we need to rely on static arrays in some areas, in debug builds there is an assert to guarantee we are not going to run out of space and MAXVAR is easy to change.